File: src/main/java/com/superfriend/superfriend/service/PermissionService.java
package com.superfriend.superfriend.service;

import com.superfriend.superfriend.dto.*;
import com.superfriend.superfriend.entity.AgentPermissionPolicy;
import com.superfriend.superfriend.mapper.AgentPermissionPolicyMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;

@Slf4j
@Service
public class PermissionService {

    @Autowired
    private AgentPermissionPolicyMapper policyMapper;

    private static final Set<String> READ_TOOL_PATTERNS = new HashSet<>(Arrays.asList(
            "read_file", "list_directory", "get_", "search_", "query_",
            "get_system_info", "list_files", "read_graph", "search_nodes"
    ));

    private static final Set<String> WRITE_TOOL_PATTERNS = new HashSet<>(Arrays.asList(
            "write_file", "create_", "save_", "upload_", "send_",
            "create_entities", "create_relations", "write_file"
    ));

    private static final Set<String> DELETE_TOOL_PATTERNS = new HashSet<>(Arrays.asList(
            "delete_", "remove_", "batch_delete", "rm_"
    ));

    private static final Set<String> EXECUTE_TOOL_PATTERNS = new HashSet<>(Arrays.asList(
            "run_command", "execute_", "shell_", "navigate_", "click_",
            "install_", "kill_", "browser"
    ));

    public PermissionCheckResult checkPermission(Long userId, String toolName, Map<String, Object> arguments) {
        String operationType = inferOperationType(toolName);

        Long effectiveUserId = userId != null ? userId : 0L;
        List<AgentPermissionPolicy> policies = policyMapper.findMatchingPolicies(effectiveUserId, toolName, operationType);

        if (policies == null || policies.isEmpty()) {
            log.debug("No matching policy for tool={}, operation={}, userId={}, defaulting to auto_allow",
                    toolName, operationType, userId);
            return PermissionCheckResult.autoAllow("默认允许所有操作");
        }

        for (AgentPermissionPolicy policy : policies) {
            String level = policy.getPermissionLevel();

            if ("auto_allow".equals(level)) {
                if (matchesResourcePattern(policy, arguments)) {
                    log.info("Permission auto_allowed: tool={}, policy={}", toolName, policy.getId());
                    PermissionCheckResult result = PermissionCheckResult.autoAllow(
                            policy.getDescription() != null ? policy.getDescription() : "策略自动允许");
                    result.setMatchedPolicyId(String.valueOf(policy.getId()));
                    return result;
                }
            } else if ("deny".equals(level)) {
                if (matchesResourcePattern(policy, arguments)) {
                    log.warn("Permission denied: tool={}, policy={}", toolName, policy.getId());
                    PermissionCheckResult result = PermissionCheckResult.denied(
                            policy.getDescription() != null ? policy.getDescription() : "策略拒绝");
                    result.setMatchedPolicyId(String.valueOf(policy.getId()));
                    return result;
                }
            }
        }

        log.debug("Permission auto_allowed by default: tool={}, operation={}", toolName, operationType);
        return PermissionCheckResult.autoAllow("默认允许所有操作");
    }

    public String inferOperationType(String toolName) {
        if (toolName == null) return "execute";

        String lowerName = toolName.toLowerCase();

        for (String pattern : DELETE_TOOL_PATTERNS) {
            if (lowerName.contains(pattern)) return "delete";
        }
        for (String pattern : WRITE_TOOL_PATTERNS) {
            if (lowerName.contains(pattern)) return "write";
        }
        for (String pattern : READ_TOOL_PATTERNS) {
            if (lowerName.contains(pattern)) return "read";
        }
        for (String pattern : EXECUTE_TOOL_PATTERNS) {
            if (lowerName.contains(pattern)) return "execute";
        }

        return "execute";
    }

    private boolean matchesResourcePattern(AgentPermissionPolicy policy, Map<String, Object> arguments) {
        String pattern = policy.getResourcePattern();
        if (pattern == null || pattern.trim().isEmpty()) {
            return true;
        }

        if (arguments == null || arguments.isEmpty()) {
            return true;
        }

        for (Object value : arguments.values()) {
            if (value != null && matchesGlobPattern(pattern, value.toString())) {
                return true;
            }
        }

        return false;
    }

    private boolean matchesGlobPattern(String pattern, String text) {
        if (pattern == null || text == null) return false;

        String regex = pattern
                .replace("\\", "\\\\")
                .replace(".", "\\.")
                .replace("*", ".*")
                .replace("?", ".");

        try {
            return text.matches(regex);
        } catch (Exception e) {
            return text.contains(pattern.replace("*", "").replace("?", ""));
        }
    }

    public List<AgentPermissionPolicy> getUserPolicies(Long userId) {
        return policyMapper.findByUserId(userId);
    }

    public AgentPermissionPolicy addPolicy(PermissionPolicyDTO dto) {
        AgentPermissionPolicy policy = new AgentPermissionPolicy();
        policy.setUserId(dto.getUserId());
        policy.setToolName(dto.getToolName());
        policy.setOperationType(dto.getOperationType());
        policy.setResourcePattern(dto.getResourcePattern());
        policy.setPermissionLevel(dto.getPermissionLevel());
        policy.setDescription(dto.getDescription());
        policyMapper.insert(policy);
        log.info("Added permission policy: userId={}, tool={}, level={}", dto.getUserId(), dto.getToolName(), dto.getPermissionLevel());
        return policy;
    }

    public AgentPermissionPolicy updatePolicy(Long id, PermissionPolicyDTO dto) {
        AgentPermissionPolicy policy = policyMapper.findById(id);
        if (policy == null) return null;

        if (dto.getToolName() != null) policy.setToolName(dto.getToolName());
        if (dto.getOperationType() != null) policy.setOperationType(dto.getOperationType());
        if (dto.getResourcePattern() != null) policy.setResourcePattern(dto.getResourcePattern());
        if (dto.getPermissionLevel() != null) policy.setPermissionLevel(dto.getPermissionLevel());
        if (dto.getDescription() != null) policy.setDescription(dto.getDescription());

        policyMapper.update(policy);
        log.info("Updated permission policy: id={}", id);
        return policy;
    }

    public boolean deletePolicy(Long id) {
        int rows = policyMapper.deleteById(id);
        log.info("Deleted permission policy: id={}, rows={}", id, rows);
        return rows > 0;
    }
}

File: superfriend-py/src/superfriend/services/permission_service.py
from __future__ import annotations

import re
import time
import structlog
from dataclasses import dataclass, field
from typing import Any
from enum import Enum

logger = structlog.get_logger()


class PermissionLevel(str, Enum):
    AUTO_ALLOW = "auto_allow"
    REQUIRE_APPROVAL = "require_approval"
    DENY = "deny"


class OperationType(str, Enum):
    READ = "read"
    WRITE = "write"
    DELETE = "delete"
    EXECUTE = "execute"


READ_TOOL_PATTERNS = {"read_file", "list_directory", "get_", "search_", "query_", "get_system_info", "list_files", "read_graph", "search_nodes"}
WRITE_TOOL_PATTERNS = {"write_file", "create_", "save_", "upload_", "send_", "create_entities", "create_relations"}
DELETE_TOOL_PATTERNS = {"delete_", "remove_", "batch_delete", "rm_"}
EXECUTE_TOOL_PATTERNS = {"run_command", "execute_", "shell_", "navigate_", "click_", "install_", "kill_", "browser"}


@dataclass
class PermissionPolicy:
    id: int | None = None
    user_id: int = 0
    tool_name: str = ""
    operation_type: str = "execute"
    resource_pattern: str = ""
    permission_level: str = PermissionLevel.AUTO_ALLOW.value
    description: str = ""


@dataclass
class PermissionCheckResult:
    allowed: bool = True
    level: str = PermissionLevel.AUTO_ALLOW.value
    reason: str = ""
    matched_policy_id: int | None = None
    needs_approval: bool = False

    @classmethod
    def auto_allow(cls, reason: str = "") -> PermissionCheckResult:
        return cls(allowed=True, level=PermissionLevel.AUTO_ALLOW.value, reason=reason)

    @classmethod
    def require_approval(cls, reason: str = "") -> PermissionCheckResult:
        return cls(allowed=False, level=PermissionLevel.REQUIRE_APPROVAL.value, reason=reason, needs_approval=True)

    @classmethod
    def denied(cls, reason: str = "") -> PermissionCheckResult:
        return cls(allowed=False, level=PermissionLevel.DENY.value, reason=reason)


class PermissionService:
    def __init__(self):
        self._policies: dict[int, PermissionPolicy] = {}
        self._next_id: int = 1

    def check_permission(self, user_id: int, tool_name: str, arguments: dict[str, Any] | None = None) -> PermissionCheckResult:
        operation_type = self.infer_operation_type(tool_name)
        effective_user_id = user_id or 0

        matching = [
            p for p in self._policies.values()
            if p.user_id == effective_user_id
            and (p.tool_name == "*" or p.tool_name == tool_name or tool_name.startswith(p.tool_name))
            and (p.operation_type == "*" or p.operation_type == operation_type)
        ]

        if not matching:
            return PermissionCheckResult.auto_allow("默认允许所有操作")

        for policy in matching:
            if policy.permission_level == PermissionLevel.AUTO_ALLOW.value:
                if self._matches_resource_pattern(policy, arguments):
                    return PermissionCheckResult.auto_allow(
                        policy.description or "策略自动允许",
                        matched_policy_id=policy.id,
                    )
            elif policy.permission_level == PermissionLevel.DENY.value:
                if self._matches_resource_pattern(policy, arguments):
                    return PermissionCheckResult.denied(policy.description or "策略拒绝")
            elif policy.permission_level == PermissionLevel.REQUIRE_APPROVAL.value:
                if self._matches_resource_pattern(policy, arguments):
                    return PermissionCheckResult.require_approval(policy.description or "需要审批")

        return PermissionCheckResult.auto_allow("默认允许所有操作")

    def infer_operation_type(self, tool_name: str) -> str:
        if not tool_name:
            return OperationType.EXECUTE.value
        lower = tool_name.lower()
        for pattern in DELETE_TOOL_PATTERNS:
            if pattern in lower:
                return OperationType.DELETE.value
        for pattern in WRITE_TOOL_PATTERNS:
            if pattern in lower:
                return OperationType.WRITE.value
        for pattern in READ_TOOL_PATTERNS:
            if pattern in lower:
                return OperationType.READ.value
        for pattern in EXECUTE_TOOL_PATTERNS:
            if pattern in lower:
                return OperationType.EXECUTE.value
        return OperationType.EXECUTE.value

    def _matches_resource_pattern(self, policy: PermissionPolicy, arguments: dict[str, Any] | None) -> bool:
        pattern = policy.resource_pattern
        if not pattern or not pattern.strip():
            return True
        if not arguments:
            return True
        for value in arguments.values():
            if value is not None and self._matches_glob(pattern, str(value)):
                return True
        return False

    def _matches_glob(self, pattern: str, text: str) -> bool:
        regex = pattern.replace(".", r"\.").replace("*", ".*").replace("?", ".")
        try:
            return bool(re.fullmatch(regex, text))
        except re.error:
            return pattern.replace("*", "").replace("?", "") in text

    def add_policy(self, user_id: int, tool_name: str = "*", operation_type: str = "*",
                   resource_pattern: str = "", permission_level: str = PermissionLevel.AUTO_ALLOW.value,
                   description: str = "") -> PermissionPolicy:
        policy_id = self._next_id
        self._next_id += 1
        policy = PermissionPolicy(
            id=policy_id, user_id=user_id, tool_name=tool_name,
            operation_type=operation_type, resource_pattern=resource_pattern,
            permission_level=permission_level, description=description,
        )
        self._policies[policy_id] = policy
        logger.info("permission_policy_added", policy_id=policy_id, tool=tool_name, level=permission_level)
        return policy

    def update_policy(self, policy_id: int, **kwargs: Any) -> PermissionPolicy | None:
        policy = self._policies.get(policy_id)
        if not policy:
            return None
        for k, v in kwargs.items():
            if hasattr(policy, k) and v is not None:
                setattr(policy, k, v)
        return policy

    def delete_policy(self, policy_id: int) -> bool:
        return self._policies.pop(policy_id, None) is not None

    def get_user_policies(self, user_id: int) -> list[PermissionPolicy]:
        return [p for p in self._policies.values() if p.user_id == user_id]

    def get_all_policies(self) -> list[PermissionPolicy]:
        return list(self._policies.values())


permission_service = PermissionService()